### ISSUE TITLE
[FakeTensor] FakeTensorMode dispatch won't change generator state

### ISSUE THREAD

**Issue by shingjan** [pytorch/pytorch#115516]
### 🐛 Describe the bug

With #110413, generator now can be passed as a parameter to fake mode dispatch. But the state of the generator passed won't change after the dispatch. As in fake mode there is no data actually generated from the dispatch but it would be nice if the generator state change can be simulated. Currently there is no way to do that on the python side, e.g. there is no advance/jump/random/random64 api in torch._C.Generator. 

```
import torch
from torch._subclasses.fake_tensor import FakeTensorMode

# with torch.no_grad():  # eager changes generator state
with FakeTensorMode():  # fake mode dispatch doesn't change generator state
    g = torch.Generator()
    s = g.get_state()
    t = torch.rand(1, generator=g)
    s1 = g.get_state()
print(all(s == s1)). # eager -> False; fake mode -> True
```

### Versions

PyTorch version: 2.2.0.dev20231210+cu121
Is debug build: False
CUDA used to build PyTorch: 12.1
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.2 LTS (x86_64)
GCC version: (Ubuntu 10.5.0-1ubuntu1~22.04) 10.5.0
Clang version: 14.0.0-1ubuntu1.1
CMake version: version 3.22.1
Libc version: glibc-2.35

Python version: 3.8.16 (default, Jun 12 2023, 18:09:05)  [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-6.2.0-36-generic-x86_64-with-glibc2.17
Is CUDA available: True
CUDA runtime version: 12.1.66
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 3070
Nvidia driver version: 535.129.03
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.9.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.9.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.9.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.9.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.9.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.9.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.9.0
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                       x86_64
CPU op-mode(s):                     32-bit, 64-bit
Address sizes:                      48 bits physical, 48 bits virtual
Byte Order:                         Little Endian
CPU(s):                             24
On-line CPU(s) list:                0-23
Vendor ID:                          AuthenticAMD
Model name:                         AMD Ryzen 9 5900X 12-Core Processor
CPU family:                         25
Model:                              33
Thread(s) per core:                 2
Core(s) per socket:                 12
Socket(s):                          1
Stepping:                           0
Frequency boost:                    enabled
CPU max MHz:                        4950.1948
CPU min MHz:                        2200.0000
BogoMIPS:                           7400.30
Flags:                              fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl nonstop_tsc cpuid extd_apicid aperfmperf rapl pni pclmulqdq monitor ssse3 fma cx16 sse4_1 sse4_2 movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb cat_l3 cdp_l3 hw_pstate ssbd mba ibrs ibpb stibp vmmcall fsgsbase bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a rdseed adx smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local clzero irperf xsaveerptr rdpru wbnoinvd arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold avic v_vmsave_vmload vgif v_spec_ctrl umip pku ospke vaes vpclmulqdq rdpid overflow_recov succor smca fsrm
Virtualization:                     AMD-V
L1d cache:                          384 KiB (12 instances)
L1i cache:                          384 KiB (12 instances)
L2 cache:                           6 MiB (12 instances)
L3 cache:                           64 MiB (2 instances)
NUMA node(s):                       1
NUMA node0 CPU(s):                  0-23
Vulnerability Gather data sampling: Not affected
Vulnerability Itlb multihit:        Not affected
Vulnerability L1tf:                 Not affected
Vulnerability Mds:                  Not affected
Vulnerability Meltdown:             Not affected
Vulnerability Mmio stale data:      Not affected
Vulnerability Retbleed:             Not affected
Vulnerability Spec rstack overflow: Mitigation; safe RET, no microcode
Vulnerability Spec store bypass:    Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:           Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:           Mitigation; Retpolines, IBPB conditional, IBRS_FW, STIBP always-on, RSB filling, PBRSB-eIBRS Not affected
Vulnerability Srbds:                Not affected
Vulnerability Tsx async abort:      Not affected

Versions of relevant libraries:
[pip3] bert-pytorch==0.0.1a4
[pip3] clip-anytorch==2.5.2
[pip3] CoCa-pytorch==0.0.7
[pip3] dalle2-pytorch==1.14.2
[pip3] ema-pytorch==0.2.3
[pip3] flake8==6.0.0
[pip3] functorch==1.14.0a0+b71aa0b
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.24.1
[pip3] open-clip-torch==2.20.0
[pip3] pytorch-transformers==1.2.0
[pip3] pytorch-triton==2.1.0+bcad9dabe1
[pip3] pytorch-warmup==0.1.1
[pip3] rotary-embedding-torch==0.2.5
[pip3] torch==2.2.0.dev20231210+cu121
[pip3] torch-fidelity==0.3.0
[pip3] torch_geometric==2.4.0
[pip3] torch-struct==0.5
[pip3] torch-tensorrt==2.2.0.dev0+4e5b0f6e
[pip3] torchaudio==2.2.0.dev20231210+cu121
[pip3] torchdata==0.7.0.dev20230921
[pip3] torchmetrics==1.0.1
[pip3] torchrec-nightly==2023.7.17
[pip3] torchtext==0.16.0.dev20230921+rocm5.6
[pip3] torchvision==0.17.0.dev20231210+cu121
[pip3] vector-quantize-pytorch==1.6.30

**Comment by ezyang** [comment]
This is very reasonable. The generator state advancement is device dependent so it will be fiddly to do right.

**Comment by shingjan** [comment]
@ezyang thanks for the response! IMO the way to do it is 1) Implement a `FakeGenerator` or `Generator(device='meta')` and then for each meta registration/dispatch, we advance the fake/meta Generator. But that requires the CPUGenerator and CUDAGenerator to have the same underlying state, e.g. philox. But a cpu philox RNG seems to be less performant than mt19937 according to previous discussions. I am more than happy to fix this for Fake Tensor and would love your opinion on this.

**Comment by ezyang** [comment]
CUDA generator's state lives entirely on host so in principle you don't need a fake generator, you can just use a regular generator. Making torch.Generator subclasseable sounds like cancer, let's not do it.

**Comment by eellison** [comment]
In some cases the number of times we advance generator state is data dependent. We won't necessarily be able to model this correctly. I don't know if this is worth fixing - why do you need this ?

**Comment by shingjan** [comment]
Thanks for the response @ezyang @eellison !

For CUDA generator one can simply calculate the offset with total number of elements randomly generated. But for CPU this is a bit tricky as we also need to update the state. In theory those can all be done in python with set/get_state(), e.g. replicating part of the logic to update cpu rng state like [here](https://github.com/pytorch/pytorch/blob/848cfe8d45d2d9715c3d16b8f7be694a49875405/aten/src/ATen/core/MT19937RNGEngine.h#L174) / cuda rng state like [here](https://github.com/pytorch/pytorch/blob/7ecddaef23fe536e8e2d12b9b7f77ce7090b3916/aten/src/ATen/native/cuda/DistributionTemplates.h#L48). So probably add an api to generator state, like advance/jump()?

@eellison My use case is that I have a Fake Mode-like dispatch mode and I want the generator used under this mode/context to have the state correctly updated. Something like:
```
with Custom_Mode(pre_dispatch=True):
    g = torch.Generator()
    t1 = torch.rand(1, generator=g)
    t2 = torch.rand(1, generator=g)
    assert(all(t1!=t2))
```

For dispatch that we can shape infer on I feel we could model the generator state correctly. Should be safe to disregard any dispatch that has dynamic output shape or is input data dependent, if we can't even shape infer those. WDYT?

**Comment by ezyang** [comment]
Yes
